horovod/common/operations.cc

```diff
@@ -267,6 +389,141 @@ const Status DUPLICATE_NAME_ERROR = Status::InvalidArgument(
+        }
+        else if (tensor_table[i].request != NULL_VAL && tensor_table[i].comm == NULL_VAL)
+        {
+            bool testflag = 0;
+            MLSL::Environment::GetEnv().Test(tensor_table[i].request, &testflag);
+
+            if (testflag)
+            {
+                if (tensor_table[i].req_type == HRT_ALLGATHER && tensor_table[i].size_vec != NULL)
+                {
+                    TensorShape single_slice_shape;
+                    size_t total_first_dimension_size = 0;
+                    size_t total_size_of_other_dims = 1;
+                    for (int j = 1; j < tensor_table[i].tensor->shape().dims(); j++)
+                    {
+                        single_slice_shape.AddDim(tensor_table[i].tensor->shape().dim_size(j));
+                        total_size_of_other_dims *= tensor_table[i].tensor->shape().dim_size(j);
+                    }
+
+                    std::vector<size_t> rcounts(horovod_global.size);
+                    size_t element_bytesize = tensor_table[i].tensor->size()/tensor_table[i].tensor->shape().num_elements();
+                    for (int j = 0; j < horovod_global.size; j++)
+                    {
+                        rcounts[j] = tensor_table[i].size_vec[j] * total_size_of_other_dims * element_bytesize;
+                        total_first_dimension_size += tensor_table[i].size_vec[j];
+                    }
+
+                    // Allgather output will have shape of:
+                    // (sum of first dimension of every tensor) x (tensor slice shape)
+                    TensorShape output_shape;
+                    output_shape.AddDim((int64_t)total_first_dimension_size);
+                    output_shape.AppendShape(single_slice_shape);
+
+                    Status status = tensor_table[i].context->AllocateOutput(output_shape, &(tensor_table[i].output));
+                    if (!status.ok())
+                    {
+                        tensor_table[i].callback(status);
+                        return;
+                    }
+
+                    tensor_table[i].request = tensor_table[i].tmp_comm->AllGatherv((void*) tensor_table[i].tensor->data(), tensor_table[i].tensor->size(),
+                                                            (void*) tensor_table[i].output->data(), rcounts.data(),
+                                                            MLSL::DT_BYTE, MLSL::GT_DATA);
+
+                    delete tensor_table[i].size_vec;
+                    tensor_table[i].size_vec = NULL;
+                    tensor_table[i].comm = NULL_VAL;
+                    MLSL_LOG(DEBUG, "Test success for intermediate AllGather call %s %p\n",
+                             tensor_table[i].tensor_name.c_str(), tensor_table[i].tensor->data());
+                }
+                else
+                {
+                    // Remove request from request_table and execute tensor callback.
+                    tensor_table[i].callback(Status::OK());
```

Comment:
Should it be a failure?

Reply:
@alsrgv I am sorry, but could you please, clarify a little bit, why do you think it should be a failure?